osdk/src/bundle/file.rs

```diff
@@ -1,66 +1,64 @@
 // SPDX-License-Identifier: MPL-2.0
 
 use std::{
-    fs, io,
+    fs,
     path::{Path, PathBuf},
+    time::SystemTime,
 };
 
-use sha2::{Digest, Sha256};
-
-/// A trait for files in a bundle. The file in a bundle should have it's digest and be validatable.
+/// A trait for files in a bundle. The file in a bundle should have its modified time and be validatable.
 pub trait BundleFile {
     fn path(&self) -> &PathBuf;
 
-    fn sha256sum(&self) -> &String;
+    fn modified_time(&self) -> &SystemTime;
 
-    fn calculate_sha256sum(&self) -> String {
-        let mut file = fs::File::open(self.path()).unwrap();
-        let mut hasher = Sha256::new();
-        let _n = io::copy(&mut file, &mut hasher).unwrap();
-        format!("{:x}", hasher.finalize())
+    fn get_modified_time(&self) -> SystemTime {
+        self.path().metadata().unwrap().modified().unwrap()
     }
 
     fn validate(&self) -> bool {
-        self.sha256sum() == &self.calculate_sha256sum()
+        self.modified_time() >= &self.get_modified_time()
     }
 }
 
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Initramfs {
     path: PathBuf,
-    sha256sum: String,
+    modified_time: SystemTime,
 }
 
 impl BundleFile for Initramfs {
     fn path(&self) -> &PathBuf {
         &self.path
     }
 
-    fn sha256sum(&self) -> &String {
-        &self.sha256sum
+    fn modified_time(&self) -> &SystemTime {
+        &self.modified_time
     }
 }
 
 impl Initramfs {
     pub fn new(path: impl AsRef<Path>) -> Self {
         let created = Self {
             path: path.as_ref().to_path_buf(),
-            sha256sum: String::new(),
+            modified_time: SystemTime::UNIX_EPOCH,
         };
         Self {
-            sha256sum: created.calculate_sha256sum(),
+            modified_time: created.get_modified_time(),
             ..created
         }
     }
 
-    /// Move the initramfs to the `base` directory and convert the path to a relative path.
+    /// Copy the initramfs to the `base` directory and convert the path to a relative path.
     pub fn copy_to(self, base: impl AsRef<Path>) -> Self {
         let name = self.path.file_name().unwrap();
         let dest = base.as_ref().join(name);
-        fs::copy(&self.path, dest).unwrap();
+        // Do not use hard_link/symlink here, otherwise `cargo build` will fail when initramfs is updated.
+        info!("Copying {:?} -> {:?}", &self.path, &dest);
+        fs::copy(&self.path, &dest).unwrap();
```

Comment:
It seems to be caused by not restoring the directory if validation fails. It should be fixed:

https://github.com/asterinas/asterinas/blob/00e3688aa8945e09a48e85013ab73900490db7ad/osdk/src/bundle/mod.rs#L87-L106